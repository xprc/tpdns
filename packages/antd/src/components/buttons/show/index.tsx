import React from "react";
import { Button } from "antd";
import { EyeOutlined } from "@ant-design/icons";
import {
    useCan,
    useNavigation,
    useTranslate,
    useResource,
    useRouterContext,
} from "@pankod/refine-core";
import { RefineButtonTestIds } from "@pankod/refine-ui-types";

import { ShowButtonProps } from "../types";

/**
 * `<ShowButton>` uses Ant Design's {@link https://ant.design/components/button/ `<Button>`} component.
 * It uses the {@link https://refine.dev/docs/core/hooks/navigation/useNavigation#show `show`} method from {@link https://refine.dev/docs/core/hooks/navigation/useNavigation `useNavigation`} under the hood.
 * It can be useful when redirecting the app to the show page with the record id route of resource.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/buttons/show-button} for more details.
 */
export const ShowButton: React.FC<ShowButtonProps> = ({
    resourceName: propResourceName,
    resourceNameOrRouteName: propResourceNameOrRouteName,
    recordItemId,
    hideText = false,
    accessControl,
    ignoreAccessControlProvider = false,
    children,
    onClick,
    ...rest
}) => {
    const accessControlEnabled =
        accessControl?.enabled ?? !ignoreAccessControlProvider;
    const hideIfUnauthorized = accessControl?.hideIfUnauthorized ?? false;
    const { showUrl: generateShowUrl } = useNavigation();
    const { Link } = useRouterContext();

    const translate = useTranslate();

    const { resourceName, id, resource } = useResource({
        resourceName: propResourceName,
        resourceNameOrRouteName: propResourceNameOrRouteName,
        recordItemId,
    });

    const { data } = useCan({
        resource: resourceName,
        action: "show",
        params: { id, resource },
        queryOptions: {
            enabled: accessControlEnabled,
        },
    });

    const createButtonDisabledTitle = () => {
        if (data?.can) return "";
        else if (data?.reason) return data.reason;
        else
            return translate(
                "buttons.notAccessTitle",
                "You don't have permission to access",
            );
    };

    const showUrl = generateShowUrl(propResourceName ?? resource.route!, id!);

    if (accessControlEnabled && hideIfUnauthorized && !data?.can) {
        return null;
    }

    return (
        <Link
            to={showUrl}
            replace={false}
            onClick={(e: React.MouseEvent<HTMLElement, MouseEvent>) => {
                if (data?.can === false) {
                    e.preventDefault();
                    return;
                }
                if (onClick) {
                    e.preventDefault();
                    onClick(e);
                }
            }}
        >
            <Button
                icon={<EyeOutlined />}
                disabled={data?.can === false}
                title={createButtonDisabledTitle()}
                data-testid={RefineButtonTestIds.ShowButton}
                {...rest}
            >
                {!hideText && (children ?? translate("buttons.show", "Show"))}
            </Button>
        </Link>
    );
};
