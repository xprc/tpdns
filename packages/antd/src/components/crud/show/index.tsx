import React from "react";
import { Card, PageHeader, Space, Spin } from "antd";
import {
    useNavigation,
    useResourceWithRoute,
    useRouterContext,
    useTranslate,
    ResourceRouterParams,
    userFriendlyResourceName,
    useRefineContext,
} from "@pankod/refine-core";

import {
    EditButton,
    DeleteButton,
    RefreshButton,
    ListButton,
    Breadcrumb,
} from "@components";
import { ShowProps } from "../types";

/**
 * `<Show>` provides us a layout for displaying the page.
 * It does not contain any logic but adds extra functionalities like a refresh button.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/basic-views/show} for more details.
 */
export const Show: React.FC<ShowProps> = ({
    title,
    canEdit,
    canDelete,
    actionButtons,
    isLoading = false,
    children,
    pageHeaderProps,
    resource: resourceFromProps,
    recordItemId,
    dataProviderName,
    breadcrumb: breadcrumbFromProps,
    contentProps,
    headerProps,
    wrapperProps,
    headerButtons,
    footerButtons,
    footerButtonProps,
    headerButtonProps,
    goBack: goBackFromProps,
}) => {
    const translate = useTranslate();

    const { goBack, list } = useNavigation();

    const resourceWithRoute = useResourceWithRoute();

    const { useParams } = useRouterContext();

    const {
        resource: routeResourceName,
        action: routeFromAction,
        id: idFromRoute,
    } = useParams<ResourceRouterParams>();

    const resource = resourceWithRoute(resourceFromProps ?? routeResourceName);

    const isDeleteButtonVisible = canDelete ?? resource.canDelete;
    const isEditButtonVisible = canEdit ?? resource.canEdit;

    const { options } = useRefineContext();
    const breadcrumb =
        typeof breadcrumbFromProps === "undefined"
            ? options?.breadcrumb
            : breadcrumbFromProps;

    const id = recordItemId ?? idFromRoute;

    const defaultHeaderButtons = (
        <>
            {!recordItemId && (
                <ListButton resourceNameOrRouteName={resource.route} />
            )}
            {isEditButtonVisible && (
                <EditButton
                    {...(isLoading ? { disabled: true } : {})}
                    resourceNameOrRouteName={resource.route}
                    recordItemId={id}
                />
            )}
            {isDeleteButtonVisible && (
                <DeleteButton
                    {...(isLoading ? { disabled: true } : {})}
                    resourceNameOrRouteName={resource.route}
                    recordItemId={id}
                    onSuccess={() => list(resource.route ?? resource.name)}
                    dataProviderName={dataProviderName}
                />
            )}
            <RefreshButton
                {...(isLoading ? { disabled: true } : {})}
                resourceNameOrRouteName={resource.route}
                recordItemId={id}
                dataProviderName={dataProviderName}
            />
        </>
    );

    return (
        <div {...(wrapperProps ?? {})}>
            <PageHeader
                ghost={false}
                backIcon={goBackFromProps}
                onBack={routeFromAction ? goBack : undefined}
                title={
                    title ??
                    translate(
                        `${resource.name}.titles.show`,
                        `Show ${userFriendlyResourceName(
                            resource.label ?? resource.name,
                            "singular",
                        )}`,
                    )
                }
                extra={
                    <Space
                        key="extra-buttons"
                        wrap
                        {...(headerButtonProps ?? {})}
                    >
                        {headerButtons
                            ? typeof headerButtons === "function"
                                ? headerButtons({
                                      defaultButtons: defaultHeaderButtons,
                                  })
                                : headerButtons
                            : defaultHeaderButtons}
                    </Space>
                }
                breadcrumb={
                    typeof breadcrumb !== "undefined" ? (
                        <>{breadcrumb}</> ?? undefined
                    ) : (
                        <Breadcrumb />
                    )
                }
                {...(pageHeaderProps ?? {})}
                {...(headerProps ?? {})}
            >
                <Spin spinning={isLoading}>
                    <Card
                        bordered={false}
                        actions={
                            footerButtons
                                ? [
                                      <Space
                                          key="footer-buttons"
                                          wrap
                                          {...footerButtonProps}
                                      >
                                          {typeof footerButtons === "function"
                                              ? footerButtons({
                                                    defaultButtons: null,
                                                })
                                              : footerButtons}
                                      </Space>,
                                  ]
                                : actionButtons
                                ? [actionButtons]
                                : undefined
                        }
                        {...(contentProps ?? {})}
                    >
                        {children}
                    </Card>
                </Spin>
            </PageHeader>
        </div>
    );
};
