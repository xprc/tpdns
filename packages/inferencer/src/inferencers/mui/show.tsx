import * as RefineMui from "@pankod/refine-mui";

import { createInferencer } from "@/create-inferencer";
import {
    jsx,
    componentName,
    prettyString,
    accessor,
    printImports,
    noOp,
    getVariableName,
} from "@/utilities";

import { ErrorComponent } from "./error";
import { LoadingComponent } from "./loading";
import { CodeViewerComponent } from "./code-viewer";

import { InferencerResultComponent, InferField, ImportElement } from "@/types";

/**
 * @experimental This is an experimental component
 */
export const ShowInferencer: InferencerResultComponent = createInferencer({
    type: "show",
    additionalScope: [["@pankod/refine-mui", "RefineMui", RefineMui]],
    codeViewerComponent: CodeViewerComponent,
    loadingComponent: LoadingComponent,
    errorComponent: ErrorComponent,
    renderer: ({ resource, fields, isCustomPage, id }) => {
        const COMPONENT_NAME = componentName(
            resource.label ?? resource.name,
            "show",
        );
        const recordName = "record";
        const imports: Array<ImportElement> = [
            ["useShow", "@pankod/refine-core"],
            ["Show", "@pankod/refine-mui"],
            ["Typography", "@pankod/refine-mui"],
            ["Stack", "@pankod/refine-mui"],
        ];

        const relationFields: (InferField | null)[] = fields.filter(
            (field) => field?.relation && !field?.fieldable && field?.resource,
        );

        const relationHooksCode = relationFields
            .filter(Boolean)
            .map((field) => {
                if (field?.relation && !field.fieldable && field.resource) {
                    if (field.multiple) {
                        imports.push(["useMany", "@pankod/refine-core"]);

                        let ids = accessor(recordName, field.key);

                        if (field.accessor) {
                            ids = `${accessor(
                                recordName,
                                field.key,
                            )}?.map((item: any) => ${accessor(
                                "item",
                                undefined,
                                field.accessor,
                            )})`;
                        }

                        return `
                    const { data: ${getVariableName(
                        field.key,
                        "Data",
                    )}, isLoading: ${getVariableName(
                            field.key,
                            "IsLoading",
                        )} } =
                    useMany({
                        resource: "${field.resource.name}",
                        ids: ${ids} || [],
                        queryOptions: {
                            enabled: !!${recordName},
                        },
                    });
                    `;
                    }

                    imports.push(["useOne", "@pankod/refine-core"]);

                    return `
                    const { data: ${getVariableName(
                        field.key,
                        "Data",
                    )}, isLoading: ${getVariableName(
                        field.key,
                        "IsLoading",
                    )} } =
                    useOne({
                        resource: "${field.resource.name}",
                        id: ${accessor(
                            recordName,
                            field.key,
                            field.accessor,
                            false,
                        )} || "",
                        queryOptions: {
                            enabled: !!${recordName},
                        },
                    });
                `;
                }
                return undefined;
            })
            .filter(Boolean);

        const renderRelationFields = (field: InferField) => {
            if (field.relation && field.resource) {
                const variableName = getVariableName(field.key, "Data");
                const variableIsLoading = getVariableName(
                    field.key,
                    "IsLoading",
                );

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    {${variableIsLoading} ? (
                        <>
                            Loading...
                        </>
                        ) : (
                        <>
                        ${(() => {
                            if (field.relationInfer) {
                                if (field.relationInfer?.accessor) {
                                    if (
                                        Array.isArray(
                                            field.relationInfer.accessor,
                                        )
                                    ) {
                                        return `Not Handled.`;
                                    } else {
                                        const mapItemName = getVariableName(
                                            field.key,
                                        );
                                        const val = accessor(
                                            mapItemName,
                                            undefined,
                                            field.relationInfer.accessor,
                                        );
                                        return `
                                            <Stack direction="row" spacing={1}>
                                                {${variableName}?.data?.map((${mapItemName}: any) => (
                                                    <TagField key={${val}} value={${val}} />
                                                ))}
                                            </Stack>
                                        `;
                                    }
                                } else {
                                    return `Not Handled.`;
                                }
                            } else {
                                return `not-handled - relation with multiple but no resource`;
                            }
                        })()}
                        </>
                    )}
                    `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>

                    {${variableIsLoading} ? (
                        <>Loading...</>
                    ) : (
                        <>
                        ${(() => {
                            if (field.relationInfer) {
                                if (field.relationInfer?.accessor) {
                                    if (
                                        Array.isArray(
                                            field.relationInfer.accessor,
                                        )
                                    ) {
                                        return `{${accessor(
                                            `${variableName}?.data`,
                                            undefined,
                                            field.relationInfer.accessor,
                                            ' + " " + ',
                                        )}}`;
                                    } else {
                                        return `{${variableName}?.data?.${field.relationInfer.accessor}}`;
                                    }
                                } else {
                                    return `{${variableName}?.data}`;
                                }
                            } else {
                                return `{${variableName}?.data?.id}`;
                            }
                        })()}
                        </>
                    )}
                    `;
            }
            return undefined;
        };

        const textFields = (field: InferField) => {
            if (field.type === "text") {
                imports.push([
                    "TextFieldComponent as TextField",
                    "@pankod/refine-mui",
                ]);

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <TagField value={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <TextField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                    )}} />
                `;
            }
            return undefined;
        };

        const imageFields = (field: InferField) => {
            if (field.type === "image") {
                if (field.multiple) {
                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        " + ",
                    )}} />
                    `;
            }
            return undefined;
        };

        const emailFields = (field: InferField) => {
            if (field.type === "email") {
                imports.push(["EmailField", "@pankod/refine-mui"]);

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <TagField value={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }
                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <EmailField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        " + ",
                    )}} />
                `;
            }
            return undefined;
        };

        const urlFields = (field: InferField) => {
            if (field.type === "url") {
                imports.push(["UrlField", "@pankod/refine-mui"]);

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <TagField value={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }
                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <UrlField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        " + ",
                    )}} />
                `;
            }
            return undefined;
        };

        const booleanFields = (field: InferField) => {
            if (field.type === "boolean") {
                imports.push(["BooleanField", "@pankod/refine-mui"]);

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(
                        recordName,
                        field.key,
                    )}?.map((item: any, index: number) => (
                        <TagField value={${val}} key={index} />
                    ))}
                    </Stack>
                `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <BooleanField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        " && ",
                    )}} />
                `;
            }
            return undefined;
        };

        const dateFields = (field: InferField) => {
            if (field.type === "date") {
                imports.push(["DateField", "@pankod/refine-mui"]);

                if (field.multiple) {
                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <DateField value={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <DateField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        " + ' ' + ",
                    )}} />
                `;
            }
            return undefined;
        };

        const richtextFields = (field: InferField) => {
            if (field.type === "richtext") {
                imports.push(["MarkdownField", "@pankod/refine-mui"]);

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <MarkdownField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        ' + " " + ',
                    )}} />
                `;
            }

            return undefined;
        };

        const numberFields = (field: InferField) => {
            if (field.type === "number") {
                imports.push(["NumberField", "@pankod/refine-mui"]);

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor("item", undefined, field.accessor);

                    return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${accessor(recordName, field.key)}?.map((item: any) => (
                        <TagField value={${val}} key={${val}} />
                    ))}
                    </Stack>
                `;
                }

                return jsx`
                    <Typography variant="body1" fontWeight="bold">
                        ${prettyString(field.key)}
                    </Typography>
                    <NumberField value={${accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        ' + " " + ',
                    )} ?? ""} />
                `;
            }
            return undefined;
        };

        const renderedFields: Array<string | undefined> = fields.map(
            (field) => {
                switch (field?.type) {
                    case "text":
                        return textFields(field);
                    case "number":
                        return numberFields(field);
                    case "richtext":
                        return richtextFields(field);
                    case "email":
                        return emailFields(field);
                    case "image":
                        return imageFields(field);
                    case "date":
                        return dateFields(field);
                    case "boolean":
                        return booleanFields(field);
                    case "url":
                        return urlFields(field);
                    case "relation":
                        return renderRelationFields(field);
                    default:
                        return undefined;
                }
            },
        );

        noOp(imports);

        return jsx`
        ${printImports(imports)}

        export const ${COMPONENT_NAME} = () => {
            const { queryResult } = useShow(${
                isCustomPage
                    ? `{ 
                        resource: "${resource.name}", 
                        id: ${id}
                    }`
                    : ""
            });
            const { data, isLoading } = queryResult;
        
            const ${recordName} = data?.data;
        
            ${relationHooksCode}

            return (
                <Show isLoading={isLoading}>
                    <Stack gap={1}>
                        ${renderedFields.join("")}
                    </Stack>
                </Show>
            );
        };
        `;
    },
});
