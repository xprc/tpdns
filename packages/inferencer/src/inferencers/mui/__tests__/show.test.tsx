import React from "react";
import { Route, Routes } from "react-router-dom";

import { render, act, TestWrapper } from "@test";
import { ShowInferencer } from "../show";

describe("MuiShowInferencer", () => {
    it("should match the snapshot", async () => {
        const Wrapper = TestWrapper({
            routerInitialEntries: ["/posts/show/11"],
            resources: [
                {
                    name: "posts",
                    list: () => <div>list</div>,
                    show: ShowInferencer,
                },
                {
                    name: "categories",
                },
                {
                    name: "tags",
                },
                { name: "users" },
            ],
        });

        const rendering = render(
            <Wrapper>
                <Routes>
                    <Route
                        path="/:resource/show/:id"
                        element={<ShowInferencer resource="posts" />}
                    />
                </Routes>
            </Wrapper>,
        );

        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });

        const node = rendering.asFragment();

        expect(node).toMatchSnapshot();
    });
});
