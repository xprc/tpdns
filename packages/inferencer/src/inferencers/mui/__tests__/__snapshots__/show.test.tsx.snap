// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MuiShowInferencer should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-12qh39f-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Show Post
          </h5>
        </div>
        <div
          class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
        >
          <div
            class="MuiBox-root css-10egq61"
          >
            <a
              href="/posts"
              style="text-decoration: none;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-198eyrn-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                title=""
                type="button"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ListOutlinedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7zm-4 6h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
                    />
                  </svg>
                </span>
                Posts
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1xanul-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              id=":r0:"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="RefreshOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
                  />
                </svg>
              </span>
              Refresh
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="css-hwnj0i-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Id
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            11
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Title
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Nobis exercitationem officia quia est.
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Slug
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            provident-culpa-facere
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Content
          </p>
          <p>
            Non velit blanditiis optio quo quaerat nam aperiam. Odit tempore consequatur voluptatibus molestias. Est nemo aut. Fugit aliquam enim eveniet veritatis. Magni et aut. Consequatur aliquid rerum vero. Quaerat debitis enim magnam. Dolorem est delectus architecto et accusantium. Velit quia vero. Aut necessitatibus quia at consequatur soluta accusantium.
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Hit
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            920,557
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Category
          </p>
          Numquam Saepe Illo
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            User
          </p>
          Eriberto Leannon
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Status
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            draft
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Status Color
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            orange
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Created At
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            04/02/2022
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Published At
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            08/10/2021
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Image
          </p>
          <div
            class="css-w4z10b-MuiStack-root"
          >
            <img
              src="http://loremflickr.com/640/480/abstract"
              style="max-width: 200px; width: 100%; height: 200px;"
            />
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Tags
          </p>
          <div
            class="css-w4z10b-MuiStack-root"
          >
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-w66kx-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                Error Eos Et
              </span>
            </div>
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-w66kx-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                Nobis Et Ut
              </span>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Language
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            2
          </p>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-ljfa65-MuiCardActions-root"
      />
    </div>
  </div>
  <div
    class="MuiBox-root css-1x21b2v"
  >
    <div
      class="MuiBox-root css-10egq61"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-t65gl3-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            class="icon icon-tabler icon-tabler-code"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <polyline
              points="7 8 3 12 7 16"
            />
            <polyline
              points="17 8 21 12 17 16"
            />
            <line
              x1="14"
              x2="10"
              y1="4"
              y2="20"
            />
          </svg>
        </span>
        Show Code
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <a
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-ik8k1-MuiButtonBase-root-MuiButton-root"
        href="https://github.com/refinedev/refine/discussions/3046"
        tabindex="0"
        target="_blank"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            class="icon icon-tabler icon-tabler-message-circle"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <path
              d="M3 20l1.3 -3.9a9 8 0 1 1 3.4 2.9l-4.7 1"
            />
            <line
              x1="12"
              x2="12"
              y1="12"
              y2="12.01"
            />
            <line
              x1="8"
              x2="8"
              y1="12"
              y2="12.01"
            />
            <line
              x1="16"
              x2="16"
              y1="12"
              y2="12.01"
            />
          </svg>
        </span>
        Give Feedback
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
  <div
    style="height: 50px;"
  />
</DocumentFragment>
`;
